org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -379,7 +400,19 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
 	public CompletableFuture<List<? extends Location>> definition(TextDocumentPositionParams position) {
 		logInfo(">> document/definition");
 		NavigateToDefinitionHandler handler = new NavigateToDefinitionHandler(this.preferenceManager);
-		return computeAsync((cc) -> handler.definition(position, toMonitor(cc)));
+		return computeAsync((cc) -> {
+			IProgressMonitor monitor = toMonitor(cc);
+			if (TRUE.equals(System.getProperty(JAVA_LSP_JOIN_ON_COMPLETION))) {
```

Comment:
if (Boolean.getProperty(JAVA_LSP_JOIN_ON_COMPLETION))

Reply:
Actually we probably want to join here in all cases, since document/definition performance impact is less visible